### PR TITLE
Refactor KTO coordinated with DPO [a/N]: Remove encoder-decoder support

### PR DESCRIPTION
**Pull request by albertvillanova**
Refactor KTO coordinated with DPO [a/N]: Remove encoder-decoder support.

This cleanup significantly simplifies the KTO trainer and makes subsequent refactoring much easier.

Part of:
- https://github.com/huggingface/trl/issues/4786#issuecomment-3724124945

Coordinated with DPO refactoring, as discussed with @qgallouedec :
- #3906

### Key Changes

1. KTOConfig
   - Removed `is_encoder_decoder` parameter and documentation
   - Removed `max_completion_length` parameter (because it is specific to encoder-decoder models) and documentation 

2. KTOTrainer

    Initialization:
    - Added clear error message when user tries to use encoder-decoder model
    - Removed self.is_encoder_decoder attribute initialization
    - Removed self.max_completion_length attribute setup
    - Hardcoded is_encoder_decoder=False in DPODataCollatorWithPadding call

    Data Processing:
    - Simplified _process_tokens() function - removed entire encoder-decoder branch (~90 lines)
    - Kept only causal LM tokenization logic

    Model Forward Pass:
    - Simplified get_batch_logps(): removed is_encoder_decoder parameter
    - Always shift labels/logits by one position (causal LM only)
    - Updated all 4 calling sites to remove the parameter

    Reference Model Computation:
    - Simplified compute_reference_log_probs() - removed encoder-decoder branches
    - Simplified _compute_kl_logps() - removed encoder-decoder conditional
    - Simplified forward() - removed encoder-decoder model_kwargs
    - Simplified _compute_loss_liger() - removed encoder-decoder branches for hidden states

3. Error Handling:
    - Users attempting to use encoder-decoder models will now receive a clear error

4. Tests
    - Test updated
    - Remove commented encoder-decoder tests

### CODE REVIEW

**Comment by qgallouedec** [comment]
Yes, as @albertvillanova says, I advocated for this change in #3906. It is important change, so I would be interested to hear the opinions of @lewtun, @edbeeching, and @kashif, and ensure we are aligned on this decision for both KTO and DPO.

File: trl/experimental/kto/kto_trainer.py [hunk 1]
```diff
@@ -502,21 +480,11 @@ def make_inputs_require_grad(module, input, output):
...
-            max_completion_length = 128
-        if args.max_completion_length is not None and self.is_encoder_decoder:
-            max_completion_length = args.max_completion_length
-
         if data_collator is None:
             data_collator = DPODataCollatorWithPadding(
                 pad_token_id=processing_class.pad_token_id,
                 label_pad_token_id=args.label_pad_token_id,
-                is_encoder_decoder=self.is_encoder_decoder,
+                is_encoder_decoder=False,
```
Comment (qgallouedec) [inline trl/experimental/kto/kto_trainer.py]: you could probably remove this

Comment (albertvillanova) [reply trl/experimental/kto/kto_trainer.py]: Done here and in other places.

**Comment by edbeeching** [comment]
> Yes, as @albertvillanova says, I advocated for this change in #3906. It is important change, so I would be interested to hear the opinions of @lewtun, @edbeeching, and @kashif, and ensure we are aligned on this decision for both KTO and DPO.

Yes I agree it is best to streamline the repo and cut features which (presumably) are not widely used.

File: trl/experimental/kto/kto_trainer.py [hunk 2]
```diff
@@ -627,13 +593,12 @@ def make_inputs_require_grad(module, input, output):
...
             fn_kwargs = {
                 "prefix": "",
-                "is_encoder_decoder": self.is_encoder_decoder,
                 "tokenizer": self.processing_class,
                 "max_length": self.max_length,
                 "truncation_mode": self.truncation_mode,
                 "label_pad_token_id": self.label_pad_token_id,
                 "max_prompt_length": self.max_prompt_length,
-                "max_completion_length": self.max_completion_length,
+                "max_completion_length": None,
```
Comment (qgallouedec) [inline trl/experimental/kto/kto_trainer.py]:
```suggestion
```

File: tests/experimental/test_kto_trainer.py [hunk 3]
```diff
@@ -172,12 +153,12 @@ def test_tokenize_and_process_tokens(self):
 
         fn_kwargs = {
             "prefix": "",
-            "is_encoder_decoder": trainer.is_encoder_decoder,
             "tokenizer": trainer.processing_class,
             "max_length": trainer.max_length,
             "truncation_mode": trainer.truncation_mode,
             "label_pad_token_id": trainer.label_pad_token_id,
             "max_prompt_length": trainer.max_prompt_length,
+            "max_completion_length": None,
```
Comment (qgallouedec) [inline tests/experimental/test_kto_trainer.py]:
```suggestion
```

Comment (qgallouedec) [reply tests/experimental/test_kto_trainer.py]: and remove all `kwargs["max_completion_length"]` in `_process_tokens`

EDIT: ignore the above, none remain

Comment (albertvillanova) [reply tests/experimental/test_kto_trainer.py]: So I leave the code unchanged then?

Comment (albertvillanova) [reply tests/experimental/test_kto_trainer.py]: OK, done.
